src/main/java/com/cloudbees/plugins/credentials/CredentialsSelectHelper.java

```diff
@@ -596,24 +595,30 @@ public WrappedCredentialsStore(@NonNull ContextResolver resolver, @NonNull Crede
          * @throws ServletException if something goes wrong.
          */
         @RequirePOST
-        public void doAddCredentials(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
+        @Restricted(NoExternalUse.class)
+        public JSONObject doAddCredentials(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
             if (!store.isDomainsModifiable()) {
                 hudson.util.HttpResponses.status(400).generateResponse(req, rsp, null);
-                FormApply.applyResponse("window.alert('Domain is read-only')").generateResponse(req, rsp, null);
+                return new JSONObject()
+                        .accumulate("message", "Domain is read-only")
+                        .accumulate("notificationType", "ERROR");
             }
             store.checkPermission(CredentialsStoreAction.CREATE);
             JSONObject data = req.getSubmittedForm();
             String domainName = data.getString("domain");
             CredentialsStoreAction.DomainWrapper wrapper = getWrappers().get(domainName);
             if (!store.getDomains().contains(wrapper.getDomain())) {
                 hudson.util.HttpResponses.status(400).generateResponse(req, rsp, null);
-                FormApply.applyResponse("window.alert('Store does not have selected domain')")
-                        .generateResponse(req, rsp, null);
+                return new JSONObject()
+                        .accumulate("message", "Store does not have selected domain")
+                        .accumulate("notificationType", "ERROR");
             }
             store.checkPermission(CredentialsStoreAction.CREATE);
             Credentials credentials = req.bindJSON(Credentials.class, data.getJSONObject("credentials"));
             store.addCredentials(wrapper.getDomain(), credentials);
-            FormApply.applyResponse("window.credentials.refreshAll();").generateResponse(req, rsp, null);
+            return new JSONObject()
+                    .accumulate("message", "Credentials created")
```

Comment:
Well, that doesn't always result in the creation of credentials, e.g when ID already exists, it won't create anything.
You could use the boolean result of `addCredentials(...)` to adapt your message.

Reply:
> You could use the boolean result of addCredentials(...) to adapt your message.

Oh, that sounds great to me, I've missed that there's return value from `addCredentials`.

Reply:
I think it would be weird to show a notification saying "Credentials not added" and not specifying reason why. So I'm unsure at the moment how to act on this, will give it another thought.

Reply:
And also if domain doesn't exist it seems. See https://github.com/jenkinsci/credentials-plugin/blob/860724bea804fb39b0967f4d3a528600e399147c/src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java#L283 Unsure under what circumstances that path would be reachable.